common/ferrostar/src/routing_adapters/osrm/mod.rs

```diff
@@ -83,20 +87,52 @@ impl Route {
             }
         })?;
         if let Some(bbox) = linestring.bounding_rect() {
-            let geometry = linestring
+            let geometry: Vec<GeographicCoordinate> = linestring
                 .coords()
                 .map(|coord| GeographicCoordinate::from(*coord))
                 .collect();
 
+            println!("Route geometry length: {}", geometry.len());
+
             let steps = route
                 .legs
                 .iter()
                 .flat_map(|leg| {
-                    leg.steps
-                        .iter()
-                        .map(|step| RouteStep::from_osrm(step, polyline_precision))
+                    // Converts all single value annotation vectors into a single vector witih a value object.
+                    let annotations = if let Some(leg_annotation) = &leg.annotation {
+                        Some(algorithms::zip_annotations(leg_annotation.clone()))
+                    } else {
+                        None
+                    };
+
+                    // Index for the annotations slice
+                    let mut start_index: usize = 0;
+
+                    return leg.steps.iter().map(move |step| {
+                        let step_geometry =
+                            get_coordinates_from_geometry(&step.geometry, polyline_precision)?;
+
+                        // Slice the annotations for the current step. The annotations array represents segments between coordinates 
+                        // and is thus one element shorter than the geometry array. To account for this, we need to subtract 2 from 
+                        // the length of the geometry array on each step (one to account for the 0-based index, and one to account 
+                        // for the length difference).
+                        let step_index_len = step_geometry.len() - 2 as usize;
```

Comment:
This could well be the buffer overflow. Need to actually assert `- 2` is correct.